### PR TITLE
KubeLatentWorker

### PR DESCRIPTION
**Pull request by tardyp**
This is a continuation of #3467   by @Zempashi

Changes:
* [x] Remove parsing of .kube/config It is actually quite complex to parse for an initial support
* [x] Replace by usage of kubectl proxy. Parsing of config, and authentication is delegated to the go cmdline
* [x] separate config management in its own classes for easier customization
* [x] Replace the usage of Jobs by Pods. Pods can be identified by name, this is necessary to kill previous lost latent worker in case of master crash or bug.

## Contributor Checklist:

* [x] I have updated the unit tests
* [x] I have created a file in the `master/buildbot/newsfragment` directory (and read the `README.txt` in that directory)
* [x] I have updated the appropriate documentation

### CODE REVIEW

**Comment by tothandras** [comment]
Hi @tardyp :slightly_smiling_face: 

> Replace the usage of Jobs by Pods. Pods can be identified by name, this is necessary to kill previous lost latent worker in case of master crash or bug.

[labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/) are perfect for this scenario, have you considered using them to find the jobs/pods?

**Comment by tardyp** [comment]
@tothandras because names are unique identifiers. 
It is easy to delete a put by name, you just need one delete REST api, while for labels, you first need to search for it/them.

This also makes the logic more similar to docker or hyper.

File: master/buildbot/util/kubeclientservice.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+
+    class LocalPP(ProcessProtocol):
+        def __init__(self, config_loader):
+            self.config_loader = config_loader
+            self.got_output_deferred = defer.Deferred()
+            self.terminated_deferred = defer.Deferred()
+
+        def outReceived(self, data):
+            if not self.got_output_deferred.called:
+                self.got_output_deferred.callback(data)
```
Comment (p12tic) [inline master/buildbot/util/kubeclientservice.py]: Do we have any guarantees of how chunked input to `outReceived` is? We could get a single character worst case, couldn't we? I think ﻿`utils.getProcessOutputAndValue` would be better here.

File: master/buildbot/util/kubeclientservice.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+    Parsing the config and setting up SSL is complex.
+    So for now, we use kubectl proxy to load the config and connect to master.
+    This will run the kube proxy as a subprocess, and return configuration with
+    http://localhost:PORT
+    """
+    kube_ctl_proxy_cmd = ['kubectl', 'proxy']  # for tests override
+
+    class LocalPP(ProcessProtocol):
+        def __init__(self, config_loader):
+            self.config_loader = config_loader
```
Comment (p12tic) [inline master/buildbot/util/kubeclientservice.py]: config_loader is unused as far as I can see

File: master/buildbot/util/kubeclientservice.py (hunk 1 above)
Comment (p12tic) [reply master/buildbot/util/kubeclientservice.py]: Maybe we should wait for the first line break and call `got_output_deferred` only then?

File: master/buildbot/util/kubeclientservice.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+class KubeInClusterConfigLoader(KubeConfigLoaderBase):
+    kube_dir = '/var/run/secrets/kubernetes.io/serviceaccount/'
+
+    kube_namespace_file = os.path.join(kube_dir, 'namespace')
+    kube_token_file = os.path.join(kube_dir, 'token')
+    kube_cert_file = os.path.join(kube_dir, 'ca.crt')
+
+    def checkConfig(self):
+        if not os.path.exists(self.kube_dir):
+            return config.error(
```
Comment (p12tic) [inline master/buildbot/util/kubeclientservice.py]: `config.error` does not return anything

File: master/buildbot/worker/latent.py [hunk 4]
```diff
@@ -307,6 +308,9 @@ def disconnect(self):
 
     @defer.inlineCallbacks
     def stopService(self):
+        # the worker might be insubstantiating from buildWaitTimeout
+        while self.insubstantiating:
+            yield asyncSleep(0.1)
```
Comment (p12tic) [inline master/buildbot/worker/latent.py]: I think it's worth having an unit test to cover the shutdown of latent workers.

**Comment by tardyp** [comment]
hyper.sh did shutdown yesterday as per plan. I didn't have time to finish this with the quality I wanted.

I will need to merge anyway, in order to enable kube on metabbot. I will address the review comment in a later PR.

**Pull request merged**